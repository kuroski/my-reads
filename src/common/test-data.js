const testBooks = {
  books: [
    {
      title: 'Book Title 1',
      subtitle: 'Book Subtitle 1',
      description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
      authors: ['Author One', 'Author Two'],
      imageLinks: { thumbnail: 'http://img01.jpg' },
      id: 'PGR2AwAAQBAJ',
      shelf: 'wantToRead'
    },
    {
      title: 'Book Title 2',
      subtitle: 'Book Subtitle 2',
      description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit2.',
      authors: ['Mist Misterious'],
      imageLinks: { thumbnail: 'http://img02.jpg' },
      id: 'PGR2AwAAQBAJ2',
      shelf: 'read'
    },
    {
      title: 'Book Title 3',
      subtitle: 'Book Subtitle 3',
      description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit3.',
      authors: ['Lucky Guy', 'Telonious Monk', 'Rapid Fast'],
      imageLinks: { thumbnail: 'http://img03.jpg' },
      id: 'PGR2AwAAQBAJ3',
      shelf: 'wantToRead'
    }
  ]
};

export { testBooks };
