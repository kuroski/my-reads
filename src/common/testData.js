import { shelfState } from './shelfState';

const testBooks = {
  books: [
    {
      title: 'Book Title 1',
      description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
      authors: ['Author One', 'Author Two'],
      imageLinks: { thumbnail: 'http://img01.jpg' },
      id: 'PGR2AwAAQBAJ',
      shelf: shelfState.WANT_TO_READ
    },
    {
      title: 'Book Title 2',
      description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit2.',
      authors: ['Mist Misterious'],
      imageLinks: { thumbnail: 'http://img02.jpg' },
      id: 'PGR2AwAAQBAJ2',
      shelf: shelfState.READ
    },
    {
      title: 'Book Title 3',
      description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit3.',
      authors: ['Lucky Guy', 'Telonious Monk', 'Rapid Fast'],
      imageLinks: { thumbnail: 'http://img03.jpg' },
      id: 'PGR2AwAAQBAJ3',
      shelf: shelfState.WANT_TO_READ
    }
  ]
};

export { testBooks };
