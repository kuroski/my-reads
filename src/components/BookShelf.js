import React from 'react';
import PropTypes from 'prop-types';
import Book from './Book';

const BookShelf = props => {
  const renderBooks = () => {
    if (!props.books || props.books.length === 0)
      return <li className="empty-message">No books on this shelve</li>;

    return props.books.map(book => (
      <li key={book.id}>
        <Book
          id={book.id}
          title={book.title}
          authors={book.authors}
          coverImage={book.imageLinks.thumbnail}
          onMove={props.onMove}
        />
      </li>
    ));
  };

  return (
    <div className="bookshelf">
      <h2 className="bookshelf-title">{props.title}</h2>
      <div className="bookshelf-books">
        <ol className="books-grid">{renderBooks()}</ol>
      </div>
    </div>
  );
};

BookShelf.propTypes = {
  title: PropTypes.string.isRequired,
  onMove: PropTypes.func.isRequired,
  books: PropTypes.array
};

BookShelf.defaultProps = {
  books: []
};

export default BookShelf;
