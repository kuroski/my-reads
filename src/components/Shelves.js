import React from 'react';
import PropTypes from 'prop-types';
import BookShelf from './BookShelf';

const Shelves = props => {
  const renderBookSHelves = () => {
    if (!props.shelves || props.shelves.length === 0)
      return <div className="empty-message">Oops, no shelve created</div>;

    return props.shelves.map(shelve => (
      <BookShelf
        key={shelve.title}
        title={shelve.title}
        books={shelve.books}
        onMove={props.onMove}
      />
    ));
  };

  return <div>{renderBookSHelves()}</div>;
};

Shelves.propTypes = {
  shelves: PropTypes.array
};

Shelves.defaultProps = {
  shelves: []
};

export default Shelves;
