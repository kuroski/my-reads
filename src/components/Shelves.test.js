import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Shelves from './Shelves';
import BookShelf from './BookShelf';
import { testBooks, testShelves } from '../common/testData';
import { shelfState } from '../common/shelfState';

describe('Shelves Container', () => {
  let props;

  const build = () => {
    return shallow(<Shelves {...props} />);
  };

  beforeAll(() => {
    props = {
      shelves: [testShelves.shelves[0]],
      books: testBooks.books,
      onMove: jest.fn()
    };
  });

  it('renders Shelves Component correctly', () => {
    const wrapper = build();
    const ShelvesTree = renderer.create(wrapper[0]).toJSON();
    expect(ShelvesTree).toMatchSnapshot();
  });

  it('renders a BookShelf component when a shelve is passed', () => {
    const wrapper = build();
    const bookShelf = wrapper.find(BookShelf);

    expect(bookShelf).toHaveLength(1);
    _expectBookShelveRenderedCorrectly(bookShelf);
  });

  it('renders a empty shelve message when no shelves prop or empty array is passed', () => {
    props.shelves = [];
    const wrapper = build();
    const emptyMessage = wrapper.find('.empty-message');

    expect(wrapper.find(BookShelf)).toHaveLength(0);
    expect(emptyMessage).toHaveLength(1);
    expect(emptyMessage.contains('Oops, no shelve created')).toBe(true);
  });

  it('renders some BookShelves components when a list of shelves are passed', () => {
    props.shelves = testShelves.shelves;
    const wrapper = build();
    const firstBookShelf = wrapper.find(BookShelf).first();

    expect(wrapper.find(BookShelf)).toHaveLength(testShelves.shelves.length);
    _expectBookShelveRenderedCorrectly(firstBookShelf);
  });

  it('must pass a bound "onMove" prop', () => {
    const wrapper = build();
    const bookShelf = wrapper.find(BookShelf).first();

    bookShelf.prop('onMove')(bookShelf.id, shelfState.CURRENTLY_READING);
    expect(props.onMove).toHaveBeenCalledTimes(1);
    expect(props.onMove).toHaveBeenCalledWith(
      bookShelf.id,
      shelfState.CURRENTLY_READING
    );
  });

  function _expectBookShelveRenderedCorrectly(bookShelf) {
    const comparableBookShelf = testShelves.shelves[0];

    expect(bookShelf.prop('name')).toEqual(comparableBookShelf.name);
    expect(bookShelf.prop('shelveId')).toEqual(comparableBookShelf.value);
    expect(bookShelf.prop('shelves')).toEqual(props.shelves);
    expect(bookShelf.prop('books')).toEqual(props.books);
    expect(bookShelf.prop('onMove')).toEqual(props.onMove);
  }
});
