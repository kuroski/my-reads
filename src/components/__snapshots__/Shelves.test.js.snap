// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shelves Container renders Shelves Container correctly 1`] = `null`;
